text/0042-extrinsics-state-version.md

```diff
@@ -0,0 +1,102 @@
+# RFC-0042: Add Extrinsic State version to `RuntimeVersion`
+
+|                 |                                               |
+|-----------------|-----------------------------------------------|
+| **Start Date**  | 25th October 2023                             |
+| **Description** | Add extrinsic state version to RuntimeVersion |
+| **Authors**     | Vedhavyas Singareddi                          |
+
+## Summary
+
+At the moment, extrinsics root is derived using `StateVersion::V0` without the possibility to use different version
+unlike storage state version.
+It would be useful for projects like `Subspace` to make extrinsics state version be part of the `RuntimeVersion`.
+
+## Motivation
+
+Since the extrinsic state version is always `StateVersion::V0`, deriving extrinsic root requires full extrinsic data.
+For `Subspace` project, we have an enshrined rollups called `Domain` with optimistic verification and Fraud proofs are
+used to detect malicious behavior.
+One of the `Fraud proof` variant is to derive `Domain` block extrinsic root on `Subspace`'s consensus chain.
+Since `StateVersion::V0` requires full extrinsic data, we are forced to pass all the extrinsics through the Fraud proof.
+One of the main challenge here is some extrinsics could be big enough that this variant of Fraud proof may not be
+included in the Consensus block due to Block's weight restriction.
+If the extrinsic root is derived using `StateVersion::V1`, then we do not need to pass the full extrinsic data but
+rather at maximum, 32 byte of extrinsic data.
+
+## Stakeholders
+
+- Technical Fellowship, in its role of maintaining system runtimes.
+
+## Explanation
+
+In order to use project specific StateVersion for extrinsic roots, we proposed
+an [implementation](https://github.com/paritytech/polkadot-sdk/pull/1691) that introduced
+parameter to `frame_system::Config` but that unfortunately did not feel correct.
+So we would like to [propose](https://github.com/paritytech/polkadot-sdk/pull/1968) adding this change to
+the `RuntimeVersion`
+object. The state version, if introduced, will fallback
+to using `StateVersion::V0` for runtimes with `CoreApi` version <=4
+and projects using `CoreApi` >= 5 will have the ability to pick State version they would like to use.
+
+If implemented, the new `RuntimeVersion` definition would look something similar to
+
+```rust
+/// Runtime version (Rococo).
+#[sp_version::runtime_version]
+pub const VERSION: RuntimeVersion = RuntimeVersion {
+		spec_name: create_runtime_str!("rococo"),
+		impl_name: create_runtime_str!("parity-rococo-v2.0"),
+		authoring_version: 0,
+		spec_version: 10020,
+		impl_version: 0,
+		apis: RUNTIME_API_VERSIONS,
+		transaction_version: 22,
+		state_version: 1,
+		extrinsic_state_version: 0,
```

Comment:
Maybe we should not add this here and instead only bump the `CoreApi` version to `5`. This should signal that the `state_version` is also used for the extrinsic root calculation.

Reply:
Since `state_version` is using V1, when runtime is updated, all the projects will move from V0 to V1 for extrinsic derivation. I'm not sure if that is okay from existing projects but I don't strong opinion on this actually. I was mainly aiming to not break as much as possible

Reply:
Yeah, I also don't know what is the best way forward...

Let's wait for some other people to comment. The RFC in general looks good to me.

Reply:
IMO `state_version` should be renamed to something else (`system_version` or whatnot), and a value of `2` means that the extrinsics trie also uses trie V1.
In other words, this should be just a version number that is increased whenever we realize that there's a better way to do things, and not a way to pimp your blockchain the way you want.

Reply:
Hmm. Yeah we probably could do this. :shrug:

Reply:
One clarifying question here is if we use `system_version` to `2`
Would that mean both storage state_version and extrinsic state_version uses always `1` ?
From what I see on the polakdot runtime here - https://github.com/polkadot-fellows/runtimes/blob/main/relay/polkadot/src/lib.rs#L139
Its till uses `state_version` 0 and so it would still use `system_version` as 0 then, which is fine.

But if other existing chains wants to use extrinsic state_version `1` while keeping the storage state_version `0`, this would not be possible with this just one `system_version` and they might accidentally break their consensus.

I still think adding this extrinsic state_version would be ideal and solves the problem above.
Thoughts @bkchr ?

Reply:
Kusama will be upgraded by to `state_version: 1` with the next runtime upgrade that should be applied on monday. Then Polkadot will also follow this. Generally all chains should upgrade to this state version, especially parachains as it can improve the storage proof size, aka giving more space when building a POV. Given this, I even think that the idea of Pierre with `system_version` is a good one.